### ISSUE TITLE
There is no difference between minArrayLike and maxArrayLike

### ISSUE THREAD

**Issue by cah-andrewfitzgerald** [DiUS/pact-jvm#266]
The following two fragments generate the same pact file:
```
@Pact(consumer = "idm-client")
    public PactFragment getUsersFragment(PactDslWithProvider builder) {
        return builder
                .given("a user with an id named 'user' exists")
                .uponReceiving("get all users")
                .path("/idm/user")
                .method("GET")
                .willRespondWith()
                .status(200)
                .body(new PactDslJsonArray()
                        .maxArrayLike(5)
                        .uuid("id")
                        .stringType("userName")
                        .stringType("email")
                        .closeObject())
                .toFragment();
    }
```
emphasis ` .maxArrayLike(5)`
```
@Pact(consumer = "idm-client")
    public PactFragment getUsersFragment(PactDslWithProvider builder) {
        return builder
                .given("a user with an id named 'user' exists")
                .uponReceiving("get all users")
                .path("/idm/user")
                .method("GET")
                .willRespondWith()
                .status(200)
                .body(new PactDslJsonArray()
                        .minArrayLike(5)
                        .uuid("id")
                        .stringType("userName")
                        .stringType("email")
                        .closeObject())
                .toFragment();
    }
```
emphasis ` .minArrayLike(5)`
```
LOH001P67QYG3QD:pacts andrew.fitzgerald$ diff minArrayLike5 maxArrayLike5 
24,26c24,26
<                         "email": "nrUfIlvNMdmjSZwsNPHv",
<                         "id": "cae2db77-487a-4b6b-8e6f-9cb7411ab2b1",
<                         "userName": "rzFvcALwDaqNBDghdalp"
---
>                         "email": "DNoDFLraARDfMgKAygIA",
>                         "id": "e0f1456b-3257-490a-aff3-757b016a8d9d",
>                         "userName": "gMjLmufHxdJctMzroWRv"
```
I would expect them to generate some sort of matcher to enforce a maximum or minimum number of array entries.
```
    "metadata": {
        "pact-specification": {
            "version": "2.0.0"
        },
        "pact-jvm": {
            "version": "3.2.7"
        }
    }
```

Related Google Groups discussion: https://groups.google.com/forum/#!topic/pact-support/qCrz5Wn7NoQ

**Comment by cah-andrewfitzgerald** [comment]
Tried setting the pact-specification version to 3, still no  difference or min/max matchers

``` 
    "interactions": [
        {
            "providerState": "a user with an id named 'user' exists",
            "description": "get all users",
            "request": {
                "method": "GET",
                "path": "/idm/user",
                "body": null
            },
            "response": {
                "status": 200,
                "headers": {
                    "Content-Type": "application/json; charset=UTF-8"
                },
                "body": [
                    {
                        "email": "DFqaDyzwrCTwsAbsyzEA",
                        "id": "9c30f0c3-2241-4642-ab04-88be0598e28d",
                        "userName": "uidlTRzPquPUhVAeELyK"
                    }
                ],
                "matchingRules": {
                    "$.body[*].userName": {
                        "match": "type"
                    },
                    "$.body[*].email": {
                        "match": "type"
                    },
                    "$.body[*].id": {
                        "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
                    }
                }
            }
        }
    ],
    "metadata": {
        "pact-specification": {
            "version": "3.0.0"
        },
        "pact-jvm": {
            "version": "3.2.7"
        }
    }
```

**Comment by uglyog** [comment]
Confirmed that the minimum and maximum checks have been dropped for root arrays. I'll fix that.

**Comment by uglyog** [comment]
This has been fixed, the matchers will now be returned as:

```json
              "matchingRules": {
                    "$.body[0][*].email": {
                        "match": "type"
                    },
                    "$.body[0][*].id": {
                        "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
                    },
                    "$.body[0]": {
                        "max": 5,
                        "match": "type"
                    },
                    "$.body[0][*].userName": {
                        "match": "type"
                    }
                }
```

and 

```json
            "matchingRules": {
                    "$.body[0][*].email": {
                        "match": "type"
                    },
                    "$.body[0][*].id": {
                        "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
                    },
                    "$.body[0]": {
                        "min": 5,
                        "match": "type"
                    },
                    "$.body[0][*].userName": {
                        "match": "type"
                    }
                }
```

**Comment by uglyog** [comment]
BTW, the form that you used `new PactDslJsonArray().maxArrayLike(5)` defines an array where the first item in the array is an array with maximum size. If this is not what you meant, use the static methods on `PactDslJsonArray` (e.g., `PactDslJsonArray.arrayMaxLike(5)`), which will define an array with a maximum size.

**Comment by cah-andrewfitzgerald** [comment]
Thanks for the quick fix!

I'm still a little bit confused by the static vs instance method differences for `PactDslJsonArray`

Shouldn't these two fragments produce different JSON?

```
        DslPart body = PactDslJsonArray
                .arrayMinLike(5)
                .uuid("id")
                
                .closeObject();
        System.out.println("body.toString() = " + body.toString());
         body = new PactDslJsonArray()
                .minArrayLike(5)
                .uuid("id")
                .closeObject();
        System.out.println("body.toString() = " + body.toString());
```
```
body.toString() = [{"id":"3873d094-0f7a-441d-baa0-3468a349600c"}]
body.toString() = [{"id":"b200dae2-ae2b-4ef3-ba3f-eea689acfe6a"}]
```

**Comment by uglyog** [comment]
You're missing the final `.closeArray()` on the instance example, if you call that you get different values:

```groovy
groovy:000> staticPart = PactDslJsonArray.arrayMinLike(5).uuid("id").closeObject()
===> [{"id":"99288281-633c-4dd9-95a7-17a8344cc090"}]
groovy:000> instancePart = new PactDslJsonArray().minArrayLike(5).uuid("id").closeObject().closeArray()
===> [[{"id":"a59df1b4-fc4b-4bd9-b508-ff5b2bdbb6d2"}]]
```

This is because ` new PactDslJsonArray()` creates a new array, then `minArrayLike(5)` creates a new array, so you have 2 levels of arrays. Hence the need for the `.closeArray()`, otherwise you're working with the inner array which is the same as the static one.

As part of the fix I created a `close()` method which will close off the object graph correctly, so if you use that instead you get the correct results:

```groovy
groovy:000> staticPart = PactDslJsonArray.arrayMinLike(5).uuid("id").close()
===> [{"id":"8eae5947-aefa-4676-a927-13ee969d182e"}]
groovy:000> instancePart = new PactDslJsonArray().minArrayLike(5).uuid("id").close()
===> [[{"id":"9add5af1-82e8-4e4e-82fb-a8205fb4ed0f"}]]
```

The static method is just a convenience method to setup the matcher correctly on the outer array. But I admit that the naming is really confusing. I'll try think of a better name for the static methods.

**Comment by cah-andrewfitzgerald** [comment]
Cool, makes sense. Thanks for the explanation!

**Comment by uglyog** [comment]
Fix has been released with 3.2.9

**Comment by cah-andrewfitzgerald** [comment]
looks good, thanks!
